File: macos/Sources/xml-macker/StructuralDiff.swift
// StructuralDiff.swift
//
// Element-aware alignment for the Diff window. A direct port of the
// Windows edition's StructuralDiff, so both apps compare files the same
// way.
//
// Why it exists: a plain line-by-line comparison of two GCAM files falls
// apart the moment the same sectors appear in a different ORDER. CORE
// lists trn_aviation_intl first under USA, an SSP override lists
// trn_freight first, and the line algorithm can only say "1,146 lines
// removed here, 1,146 lines added over there", which reads as if the
// second file had not loaded at all.
//
// This walker pairs elements by what they ARE, the tag name plus the
// identifying attribute (name, year, type, id, key), wherever they sit
// among their siblings. It recurses into every matched pair and only
// line-compares the text INSIDE matched elements. An element with no
// partner becomes one clean "only in this file" block.
//
// The output is the same [DiffEngine.Hunk] the line engine produces, so
// navigation, copying and the aligned views work unchanged. Left lines
// are emitted in strict file order, which the tree sidebar depends on,
// and the walker checks that it held. If anything about a file breaks
// the assumptions (elements sharing lines, a range that does not
// partition cleanly), align returns nil and the caller falls back to the
// line comparison. Never a broken model.

import Foundation

enum StructuralDiff {

    /// One run of aligned rows. Line indices are 0-based; a hunk may have
    /// zero lines on one side.
    ///
    /// The equal runs are explicit rather than implied by the gaps
    /// between hunks, because pairing by identity visits the RIGHT file
    /// out of order: when two regions swap places, the run of equal lines
    /// after the first difference is at line 4 on the left and line 1 on
    /// the right. A hunk list with implicit gaps cannot say that.
    struct Segment {
        let isHunk: Bool
        var leftStart: Int, leftCount: Int, rightStart: Int, rightCount: Int
        static func same(_ l: Int, _ r: Int, _ count: Int) -> Segment {
            Segment(isHunk: false, leftStart: l, leftCount: count, rightStart: r, rightCount: count)
        }
        static func diff(_ l: Int, _ lc: Int, _ r: Int, _ rc: Int) -> Segment {
            Segment(isHunk: true, leftStart: l, leftCount: lc, rightStart: r, rightCount: rc)
        }
    }

    /// Attribute names that identify an element to a reader, in priority
    /// order.
    private static let keyNames = ["name", "year", "type", "id", "key"]

    /// Aligns two parsed documents. Returns the segment list in display
    /// order, or nil when the files do not fit the element-per-line shape
    /// this walker relies on.
    static func align(leftRoot: XMLTreeNode, rightRoot: XMLTreeNode,
                      leftLines: [String], rightLines: [String]) -> [Segment]? {
        let w = Walker(leftLines, rightLines)
        w.emitPair(leftRoot, 0, leftLines.count, rightRoot, 0, rightLines.count, 0)
        guard w.valid, w.leftNext == leftLines.count else { return nil }

        // Every line of BOTH files must be shown exactly once, or the two
        // sides would not line up on screen. Prove it before returning.
        var lSeen = 0, rSeen = 0
        for s in w.segments { lSeen += s.leftCount; rSeen += s.rightCount }
        guard lSeen == leftLines.count, rSeen == rightLines.count else { return nil }
        return w.segments
    }

    /// The same shape for the line engine, so the model builder has one
    /// code path: the equal runs between hunks made explicit.
    static func segments(fromLineHunks hunks: [DiffEngine.Hunk],
                         leftCount: Int, rightCount: Int) -> [Segment] {
        var out: [Segment] = []
        var li = 0, ri = 0
        for h in hunks {
            let gap = h.leftStart - li
            if gap > 0 { out.append(.same(li, ri, gap)) }
            li = h.leftStart; ri = h.rightStart
            out.append(.diff(h.leftStart, h.leftCount, h.rightStart, h.rightCount))
            li += h.leftCount; ri += h.rightCount
        }
        if leftCount - li > 0 { out.append(.same(li, ri, leftCount - li)) }
        return out
    }

    private final class Walker {
        private static let maxDepth = 80

        private let l: [String]
        private let r: [String]
        var segments: [Segment] = []

        /// Invariant: every left line is emitted exactly once, in order.
        var leftNext = 0
        var valid = true

        /// Where an unpaired block would be inserted on the other side:
        /// right after the last content shown on that side.
        private var lCursor = 0
        private var rCursor = 0

        init(_ l: [String], _ r: [String]) { self.l = l; self.r = r }

        // MARK: emitting

        private func same(_ l: Int, _ r: Int, _ count: Int) {
            if count <= 0 { return }
            if l != leftNext { valid = false; return }
            segments.append(.same(l, r, count))
            leftNext = l + count
            lCursor = l + count
            rCursor = r + count
        }

        private func diff(_ l: Int, _ lc: Int, _ r: Int, _ rc: Int) {
            if lc <= 0 && rc <= 0 { return }
            if lc > 0 && l != leftNext { valid = false; return }

            // A one-sided block that continues straight on from the
            // previous hunk on the same side, with the other side
            // unmoved, joins it. Five unmatched siblings in a row are
            // then ONE difference, so the count stops being a count of
            // elements.
            if let prev = segments.last, prev.isHunk {
                let prevLeftEnd = prev.leftStart + prev.leftCount == l
                let prevRightEnd = prev.rightStart + prev.rightCount == r
                let joinLeft = lc > 0 && rc == 0 && prevLeftEnd && prevRightEnd
                let joinRight = rc > 0 && lc == 0 && prevRightEnd && prevLeftEnd
                if joinLeft || joinRight {
                    segments[segments.count - 1] = .diff(prev.leftStart, prev.leftCount + lc,
                                                         prev.rightStart, prev.rightCount + rc)
                    if lc > 0 { leftNext = l + lc; lCursor = l + lc }
                    if rc > 0 { rCursor = r + rc }
                    return
                }
            }

            segments.append(.diff(l, lc, r, rc))
            if lc > 0 { leftNext = l + lc; lCursor = l + lc }
            if rc > 0 { rCursor = r + rc }
        }

        /// Line-compares two ranges with the patience engine and emits
        /// the result.
        private func emitLineDiff(_ lFrom: Int, _ lTo: Int, _ rFrom: Int, _ rTo: Int) {
            let lc = lTo - lFrom, rc = rTo - rFrom
            if lc <= 0 && rc <= 0 { return }
            if lc <= 0 || rc <= 0 { diff(lFrom, lc, rFrom, rc); return }

            let hunks = DiffEngine.diffSlice(left: l, lFrom: lFrom, lTo: lTo,
                                             right: r, rFrom: rFrom, rTo: rTo)
            var li = lFrom, ri = rFrom
            for h in hunks {
                let equal = h.leftStart - li
                if equal != h.rightStart - ri { valid = false; return }   // engine invariant
                same(li, ri, equal)
                diff(h.leftStart, h.leftCount, h.rightStart, h.rightCount)
                li = h.leftStart + h.leftCount
                ri = h.rightStart + h.rightCount
                if !valid { return }
            }
            if lTo - li != rTo - ri { valid = false; return }
            same(li, ri, lTo - li)
        }

        // MARK: the structural walk

        private static func kids(_ n: XMLTreeNode) -> [XMLTreeNode] {
            n.children.filter { $0.kind == .element }
        }

        private static func key(_ n: XMLTreeNode) -> String? {
            for name in keyNames {
                for a in n.attributes where a.name == name { return a.value }
            }
            return nil
        }

        /// Cuts a node's line range into: its opening line(s), one block
        /// per element child (each block runs from the child's first line
        /// up to the next child's first line, so comment and text lines
        /// between children travel with the child before them), and its
        /// closing line(s). Returns nil when the children do not sit on
        /// distinct, ascending lines.
        private static func partition(_ kids: [XMLTreeNode], _ from: Int, _ to: Int)
            -> (openEnd: Int, blockStart: [Int], closeStart: Int)? {
            guard !kids.isEmpty else { return nil }
            let first = kids[0].startLine - 1              // 0-based
            let lastEnd = kids[kids.count - 1].endLine     // 0-based EXCLUSIVE end
            guard first >= from, lastEnd <= to else { return nil }

            var blockStart = [Int](repeating: 0, count: kids.count + 1)
            for i in 0..<kids.count {
                let s = kids[i].startLine - 1
                let e = kids[i].endLine                    // exclusive
                if e < s + 1 { return nil }
                if i > 0, s < blockStart[i - 1] + 1 { return nil }        // must move forward
                if i > 0, kids[i - 1].endLine > s { return nil }          // previous child must have closed
                blockStart[i] = s
            }
            blockStart[kids.count] = lastEnd
            return (first, blockStart, lastEnd)
        }

        func emitPair(_ lNode: XMLTreeNode, _ lFrom: Int, _ lTo: Int,
                      _ rNode: XMLTreeNode, _ rFrom: Int, _ rTo: Int, _ depth: Int) {
            if !valid { return }

            let lKids = Self.kids(lNode)
            let rKids = Self.kids(rNode)

            guard depth < Self.maxDepth, !lKids.isEmpty, !rKids.isEmpty,
                  let lPart = Self.partition(lKids, lFrom, lTo),
                  let rPart = Self.partition(rKids, rFrom, rTo) else {
                emitLineDiff(lFrom, lTo, rFrom, rTo)
                return
            }
            let lBlock = lPart.blockStart, rBlock = rPart.blockStart

            // 1. The opening tag line(s).
            emitLineDiff(lFrom, lPart.openEnd, rFrom, rPart.openEnd)
            if !valid { return }

            // 2. Pair the children by identity, whatever their order.
            struct Identity: Hashable { let name: String; let key: String? }
            var rQueue: [Identity: [Int]] = [:]
            for j in 0..<rKids.count {
                let k = Identity(name: rKids[j].name, key: Self.key(rKids[j]))
                rQueue[k, default: []].append(j)
            }
            var match = [Int](repeating: -1, count: lKids.count)
            var rMatched = [Bool](repeating: false, count: rKids.count)
            for i in 0..<lKids.count {
                let k = Identity(name: lKids[i].name, key: Self.key(lKids[i]))
                if var q = rQueue[k], !q.isEmpty {
                    let j = q.removeFirst()
                    rQueue[k] = q
                    match[i] = j
                    rMatched[j] = true
                }
            }

            // A right-only child is shown just before the next right child
            // that DID find a partner, so it appears where it sits in its
            // own file; leftovers after the last partner go at the end.
            var before: [Int: [Int]] = [:]
            var pending: [Int] = []
            for j in 0..<rKids.count {
                if !rMatched[j] { pending.append(j); continue }
                if !pending.isEmpty { before[j] = pending; pending = [] }
            }
            let tail = pending

            // 3. Walk the LEFT children in file order.
            for i in 0..<lKids.count {
                let j = match[i]
                if j >= 0 {
                    if let extras = before[j] {
                        for jj in extras { emitRightOnly(rBlock[jj], rBlock[jj + 1]) }
                    }
                    emitPair(lKids[i], lBlock[i], lBlock[i + 1],
                             rKids[j], rBlock[j], rBlock[j + 1], depth + 1)
                } else {
                    emitLeftOnly(lBlock[i], lBlock[i + 1])
                }
                if !valid { return }
            }
            for jj in tail { emitRightOnly(rBlock[jj], rBlock[jj + 1]) }

            // 4. The closing tag line(s).
            emitLineDiff(lPart.closeStart, lTo, rPart.closeStart, rTo)
        }

        private func emitLeftOnly(_ from: Int, _ to: Int) { diff(from, to - from, rCursor, 0) }
        private func emitRightOnly(_ from: Int, _ to: Int) { diff(lCursor, 0, from, to - from) }
    }
}

File: windows/src/xml-macker/Windows/StructuralDiff.cs
using System.Collections.Generic;
using XMLMacker.Core;

namespace XMLMacker.Windows;

/// <summary>
/// Element-aware alignment for the Diff window.
///
/// <para>A plain line-by-line comparison of two GCAM files falls apart the moment the same sectors appear
/// in a different ORDER: CORE lists <c>trn_aviation_intl</c> first under USA, an SSP override lists
/// <c>trn_freight</c> first, and the line algorithm can only report "1,146 lines removed here, 1,146
/// lines added over there". This walker instead pairs elements by what they ARE, the tag name plus the
/// identifying attribute (<c>name</c>, <c>year</c>, <c>type</c>, <c>id</c>, <c>key</c>), regardless of
/// where they sit among their siblings, recurses into every matched pair, and only line-compares the
/// text INSIDE matched elements. Elements with no partner become one clean "only in this file" block.</para>
///
/// <para>Output is a flat list of segments in display order: <c>Same</c> runs (identical lines on both
/// sides) and <c>Diff</c> hunks (the same shape the line engine produces, so navigation and Copy work
/// unchanged). Left lines are emitted in strict file order, the tree sidebar depends on that, and the
/// walker verifies it; if anything about a file breaks the assumptions (elements sharing lines, a range
/// that does not partition cleanly), <see cref="Align"/> returns <c>null</c> and the caller falls back to
/// the line comparison. Never a broken model.</para>
/// </summary>
internal static class StructuralDiff
{
    /// <summary>One run of aligned rows. Line indices are 0-based; counts may be 0 on one side of a hunk.</summary>
    internal readonly struct Segment
    {
        public readonly bool IsHunk;
        public readonly int LeftStart, LeftCount, RightStart, RightCount;

        private Segment(bool hunk, int l, int lc, int r, int rc)
        { IsHunk = hunk; LeftStart = l; LeftCount = lc; RightStart = r; RightCount = rc; }

        public static Segment Same(int l, int r, int count) => new(false, l, count, r, count);
        public static Segment Diff(int l, int lc, int r, int rc) => new(true, l, lc, r, rc);
    }

    /// <summary>Attribute names that identify an element to a reader, in priority order.</summary>
    private static readonly string[] KeyNames = { "name", "year", "type", "id", "key" };

    /// <summary>
    /// Aligns two parsed documents. Returns the segment list, or <c>null</c> when the files do not fit the
    /// element-per-line shape this walker relies on (the caller then uses the line comparison).
    /// </summary>
    public static List<Segment>? Align(XmlTreeNode leftRoot, XmlTreeNode rightRoot,
                                       IReadOnlyList<string> leftLines, IReadOnlyList<string> rightLines)
    {
        var w = new Walker(leftLines, rightLines);
        w.EmitPair(leftRoot, 0, leftLines.Count, rightRoot, 0, rightLines.Count, 0);
        if (!w.Valid || w.LeftNext != leftLines.Count) return null;
        return w.Segments;
    }

    private sealed class Walker
    {
        private const int MaxDepth = 80;

        private readonly IReadOnlyList<string> _l;
        private readonly IReadOnlyList<string> _r;
        public readonly List<Segment> Segments = new();

        // Invariant: every left line is emitted exactly once, in order.
        public int LeftNext;
        public bool Valid = true;

        // Where an unpaired block would be inserted on the other side: right after the last content
        // shown on that side. Kept in sync by every emit.
        private int _lCursor;
        private int _rCursor;

        public Walker(IReadOnlyList<string> l, IReadOnlyList<string> r) { _l = l; _r = r; }

        // ── emitting ────────────────────────────────────────────────────────────────────────────

        private void Same(int l, int r, int count)
        {
            if (count <= 0) return;
            if (l != LeftNext) { Valid = false; return; }
            Segments.Add(Segment.Same(l, r, count));
            LeftNext = l + count;
            _lCursor = l + count;
            _rCursor = r + count;
        }

        private void Diff(int l, int lc, int r, int rc)
        {
            if (lc <= 0 && rc <= 0) return;
            if (lc > 0 && l != LeftNext) { Valid = false; return; }

            // A one-sided block that continues straight on from the previous hunk on the same side, with
            // the other side unmoved, joins it. Five unmatched siblings in a row are then ONE difference
            // ("9 lines only in CORE"), which is what the caption on the filler run already says, and the
            // difference count stops being a count of elements.
            if (Segments.Count > 0 && Segments[^1].IsHunk)
            {
                Segment prev = Segments[^1];
                bool prevLeftEnd = prev.LeftStart + prev.LeftCount == l;
                bool prevRightEnd = prev.RightStart + prev.RightCount == r;
                bool joinLeft = lc > 0 && rc == 0 && prevLeftEnd && prevRightEnd;
                bool joinRight = rc > 0 && lc == 0 && prevRightEnd && prevLeftEnd;
                if (joinLeft || joinRight)
                {
                    Segments[^1] = Segment.Diff(prev.LeftStart, prev.LeftCount + lc, prev.RightStart, prev.RightCount + rc);
                    if (lc > 0) { LeftNext = l + lc; _lCursor = l + lc; }
                    if (rc > 0) _rCursor = r + rc;
                    return;
                }
            }

            Segments.Add(Segment.Diff(l, lc, r, rc));
            if (lc > 0) { LeftNext = l + lc; _lCursor = l + lc; }
            if (rc > 0) _rCursor = r + rc;
        }

        /// <summary>Line-compares two ranges with the patience engine and emits the result.</summary>
        private void EmitLineDiff(int lFrom, int lTo, int rFrom, int rTo)
        {
            int lc = lTo - lFrom, rc = rTo - rFrom;
            if (lc <= 0 && rc <= 0) return;
            if (lc <= 0 || rc <= 0) { Diff(lFrom, lc, rFrom, rc); return; }

            List<Hunk> hunks = DiffEngine.DiffSlice(_l, lFrom, lTo, _r, rFrom, rTo);
            int li = lFrom, ri = rFrom;
            foreach (Hunk h in hunks)
            {
                int equal = h.LeftStart - li;
                if (equal != h.RightStart - ri) { Valid = false; return; }   // engine invariant
                Same(li, ri, equal);
                Diff(h.LeftStart, h.LeftCount, h.RightStart, h.RightCount);
                li = h.LeftStart + h.LeftCount;
                ri = h.RightStart + h.RightCount;
                if (!Valid) return;
            }
            if (lTo - li != rTo - ri) { Valid = false; return; }
            Same(li, ri, lTo - li);
        }

        // ── the structural walk ─────────────────────────────────────────────────────────────────

        private static List<XmlTreeNode> Kids(XmlTreeNode n)
        {
            var kids = new List<XmlTreeNode>();
            foreach (XmlTreeNode c in n.Children) if (c.Kind == NodeKind.Element) kids.Add(c);
            return kids;
        }

        private static string? Key(XmlTreeNode n)
        {
            foreach (string name in KeyNames)
                foreach ((string Name, string Value) a in n.Attributes)
                    if (a.Name == name) return a.Value;
            return null;
        }

        /// <summary>
        /// Cuts a node's line range into: its opening line(s), one block per element child (each block
        /// runs from the child's first line up to the next child's first line, so comment and text lines
        /// between children travel with the child before them), and its closing line(s).
        /// Returns false when the children do not sit on distinct, ascending lines.
        /// </summary>
        private static bool Partition(List<XmlTreeNode> kids, int from, int to,
                                      out int openEnd, out int[] blockStart, out int closeStart)
        {
            openEnd = from; closeStart = to; blockStart = new int[kids.Count + 1];
            if (kids.Count == 0) return false;

            int first = kids[0].StartLine - 1;                 // 0-based
            int lastEnd = kids[^1].EndLine;                    // 0-based EXCLUSIVE end of the last child
            if (first < from || lastEnd > to) return false;

            for (int i = 0; i < kids.Count; i++)
            {
                int s = kids[i].StartLine - 1;
                int e = kids[i].EndLine;                        // exclusive
                if (e < s + 1) return false;
                if (i > 0 && s < blockStart[i - 1] + 1) return false;      // must move forward
                if (i > 0 && kids[i - 1].EndLine > s) return false;         // previous child must have closed
                blockStart[i] = s;
            }
            blockStart[kids.Count] = lastEnd;
            openEnd = first;
            closeStart = lastEnd;
            return true;
        }

        public void EmitPair(XmlTreeNode lNode, int lFrom, int lTo, XmlTreeNode rNode, int rFrom, int rTo, int depth)
        {
            if (!Valid) return;

            List<XmlTreeNode> lKids = Kids(lNode);
            List<XmlTreeNode> rKids = Kids(rNode);

            // Pre-initialised because the compiler cannot see through the short-circuiting chain below.
            int lOpenEnd = lFrom, rOpenEnd = rFrom, lCloseStart = lTo, rCloseStart = rTo;
            int[] lBlock = System.Array.Empty<int>(), rBlock = System.Array.Empty<int>();
            bool structural = depth < MaxDepth
                && lKids.Count > 0 && rKids.Count > 0
                && Partition(lKids, lFrom, lTo, out lOpenEnd, out lBlock, out lCloseStart)
                && Partition(rKids, rFrom, rTo, out rOpenEnd, out rBlock, out rCloseStart);

            if (!structural)
            {
                EmitLineDiff(lFrom, lTo, rFrom, rTo);
                return;
            }

            // 1. The opening tag line(s).
            EmitLineDiff(lFrom, lOpenEnd, rFrom, rOpenEnd);
            if (!Valid) return;

            // 2. Pair the children by identity, whatever their order.
            var rQueue = new Dictionary<(string Name, string? Key), Queue<int>>();
            for (int j = 0; j < rKids.Count; j++)
            {
                var k = (rKids[j].Name, Key(rKids[j]));
                if (!rQueue.TryGetValue(k, out Queue<int>? q)) { q = new Queue<int>(); rQueue[k] = q; }
                q.Enqueue(j);
            }
            var match = new int[lKids.Count];
            var rMatched = new bool[rKids.Count];
            for (int i = 0; i < lKids.Count; i++)
            {
                match[i] = -1;
                var k = (lKids[i].Name, Key(lKids[i]));
                if (rQueue.TryGetValue(k, out Queue<int>? q) && q.Count > 0)
                {
                    int j = q.Dequeue();
                    match[i] = j;
                    rMatched[j] = true;
                }
            }

            // A right-only child is shown just before the next right child that DID find a partner, so it
            // appears where it sits in its own file; leftovers after the last partner go at the end.
            var before = new Dictionary<int, List<int>>();
            var pending = new List<int>();
            for (int j = 0; j < rKids.Count; j++)
            {
                if (!rMatched[j]) { pending.Add(j); continue; }
                if (pending.Count > 0) { before[j] = pending; pending = new List<int>(); }
            }
            List<int> tail = pending;

            // 3. Walk the LEFT children in file order.
            for (int i = 0; i < lKids.Count; i++)
            {
                int j = match[i];
                if (j >= 0)
                {
                    if (before.TryGetValue(j, out List<int>? extras))
                        foreach (int jj in extras) EmitRightOnly(rBlock[jj], rBlock[jj + 1]);
                    EmitPair(lKids[i], lBlock[i], lBlock[i + 1], rKids[j], rBlock[j], rBlock[j + 1], depth + 1);
                }
                else
                {
                    EmitLeftOnly(lBlock[i], lBlock[i + 1]);
                }
                if (!Valid) return;
            }
            foreach (int jj in tail) EmitRightOnly(rBlock[jj], rBlock[jj + 1]);

            // 4. The closing tag line(s).
            EmitLineDiff(lCloseStart, lTo, rCloseStart, rTo);
        }

        private void EmitLeftOnly(int from, int to) => Diff(from, to - from, _rCursor, 0);
        private void EmitRightOnly(int from, int to) => Diff(_lCursor, 0, from, to - from);
    }
}
